thorlcr/thorutil/thmem.cpp

```diff
@@ -543,39 +520,60 @@ inline const void *CThorExpandingRowArray::_allocateRowTable(rowidx_t num, unsig
     }
 }
 
-inline bool CThorExpandingRowArray::_ensure(rowidx_t requiredRows, unsigned maxSpillCost)
+inline bool CThorExpandingRowArray::doResize(void **&_rows, rowidx_t requiredRows, bool copy, unsigned maxSpillCost, memsize_t &capacity, const char *errMsg)
+{
+    //Only the writer is allowed to reallocate rows (otherwise append can't be optimized), so rows is valid outside the lock
+
+    // NB: only resize alters row capacity, so no locking required to protect getRowsCapacity()
+    rowidx_t currentMaxRows = getRowsCapacity();
+    capacity = ((memsize_t)getNewSize(requiredRows)) * sizeof(void *);
+
+    CResizeRowCallback callback(_rows, capacity, queryLock());
+    if (!resizeRowTable(_rows, capacity, copy, callback, maxSpillCost)) // callback will reset capacity
+    {
+        if (throwOnOom)
+            throw MakeActivityException(&activity, 0, "Out of memory, resizing %s, had %"RIPF"d, trying to allocate %"RIPF"d elements", errMsg, ordinality(), requiredRows);
+        return false;
+    }
+    return true;
+}
+
+inline bool CThorExpandingRowArray::_resize(rowidx_t requiredRows, unsigned maxSpillCost)
 {
     //Only the writer is allowed to reallocate rows (otherwise append can't be optimized), so rows is valid outside the lock
     if (0 == requiredRows)
+    {
+        CThorArrayLockBlock block(queryLock());
+        clearRows();
+        ReleaseThorRow(rows);
+        ReleaseThorRow(stableTable);
+        rows = NULL;
+        stableTable = NULL;
+        numRows = maxRows = 0;
         return true;
+    }
 
-    // NB: only ensure alters row capacity, so no locking required to protect getRowsCapacity()
+    // NB: only resize alters row capacity, so no locking required to protect getRowsCapacity()
     memsize_t capacity = rows ? RoxieRowCapacity(rows) : 0;
     rowidx_t currentMaxRows = getRowsCapacity();
-    if (currentMaxRows < requiredRows) // check, because may have expanded previously, but failed to allocate stableTable and set new maxRows
+    if (currentMaxRows > requiredRows) // shrink
     {
-        capacity = ((memsize_t)getNewSize(requiredRows)) * sizeof(void *);
-
-        CResizeRowCallback callback(*(void ***)(&rows), capacity, queryLock());
-        if (!resizeRowTable((void **)rows, capacity, true, callback, maxSpillCost)) // callback will reset capacity
+        if (numRows > requiredRows)
         {
-            if (throwOnOom)
-                throw MakeActivityException(&activity, 0, "Out of memory, allocating row array, had %"RIPF"d, trying to allocate %"RIPF"d elements", ordinality(), requiredRows);
-            return false;
+            for (rowidx_t i = requiredRows; i < numRows; i++)
+                ReleaseThorRow(rows[i]);
+            numRows -= (numRows-requiredRows);
         }
+        doResize((void **&)rows, requiredRows, true, maxSpillCost, capacity, "row array");
```

Comment:
You're not checking the return value from doResize, here and in the following branches.  (Returning false if it fails)